pkg/resource/deploy/deployment_executor.go

```diff
@@ -671,11 +672,15 @@ func (ex *deploymentExecutor) rebuildBaseState(resourceToStep map[*resource.Stat
 			olds[new.URN] = new
 		}
 	}
+	// Make a deep copies before mutating in undangleParentResources,
```

Comment:
This is concerning, snapshots can get fairly large (MBs to GBs) and having to deep copy them for every resource is going to result in N^N memory usage.